### PR TITLE
Access control: endpoint for searching single user permissions

### PR DESCRIPTION
**Pull request by IevaVasiljeva**
**What is this feature?**

A new endpoint (`GET /api/access-control/user/:userID/permissions/search`) for searching single user's permissions by providing an action or an action prefix.

**Why do we need this feature?**

This endpoint will be used by OnCall for the proof of concept for using RBAC with app plugins.

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer**:

I'm reusing the DB method introduced for listing all users permissions. I'm not entirely happy with doing this, as the method is pretty long and difficult to read when extra filters are introduced. However, if I go with a new method I will end up duplicating quite a lot of the code. Let me know what you think makes more sense.

Enterprise counterpart: https://github.com/grafana/grafana-enterprise/pull/4180

### CODE REVIEW

File: pkg/services/accesscontrol/acimpl/service.go [hunk 1]
```diff
@@ -335,3 +335,83 @@ func (s *Service) SearchUsersPermissions(ctx context.Context, user *user.SignedI
...
+	dbPermissions, err := s.store.SearchUsersPermissions(ctx, orgID, []int64{userID}, searchOptions)
+	if err != nil {
+		return nil, err
+	}
+	permissions = append(permissions, dbPermissions[userID]...)
+
+	return permissions, nil
+}
+
+func (s *Service) searchUserPermissionsFromCache(userID, orgID int64, searchOptions accesscontrol.SearchOptions) ([]accesscontrol.Permission, bool) {
```
Comment (gamab) [inline pkg/services/accesscontrol/acimpl/service.go:381]: can you maybe add a test to check that the cache is correctly populated and used?

**Review by gamab — APPROVED** [review]
Excellent :tada: 
I left a couple of comments for you to address, but nothing too serious :+1:

File: pkg/services/accesscontrol/acimpl/service.go [hunk 2]
```diff
@@ -335,3 +335,83 @@ func (s *Service) SearchUsersPermissions(ctx context.Context, user *user.SignedI
...
+	var roles []string
+	var ok bool
+	if roles, ok = roleList[userID]; !ok {
+		return nil, fmt.Errorf("found no basic roles for user %d in organisation %d", userID, orgID)
+	}
+	permissions := make([]accesscontrol.Permission, 0)
+	for _, builtin := range roles {
+		if basicRole, ok := s.roles[builtin]; ok {
+			for _, permission := range basicRole.Permissions {
+				if PermissionMatchesSearchOptions(permission, searchOptions) {
```
[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/accesscontrol/acimpl/service.go:368]: Yeah, I was planning to do that, but wanted to tackle it separately (and fix the wildcard issue as well).

File: pkg/services/accesscontrol/acimpl/service.go [hunk 3]
```diff
@@ -335,3 +335,83 @@ func (s *Service) SearchUsersPermissions(ctx context.Context, user *user.SignedI
 
 	return res, nil
 }
+
+func (s *Service) SearchUserPermissions(ctx context.Context, userID, orgID int64, searchOptions accesscontrol.SearchOptions) ([]accesscontrol.Permission, error) {
+	timer := prometheus.NewTimer(metrics.MAccessPermissionsSummary)
```
[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/accesscontrol/acimpl/service.go:340]: Will do (in the next PR).

File: pkg/services/accesscontrol/database/database.go [hunk 4]
```diff
@@ -73,30 +93,33 @@ func (s *AccessControlStore) SearchUsersPermissions(ctx context.Context, orgID i
 		FROM (
 			SELECT ur.user_id, ur.org_id, p.action, p.scope
 				FROM permission AS p
-				INNER JOIN user_role AS ur on ur.role_id = p.role_id
-			UNION ALL
+				INNER JOIN user_role AS ur on ur.role_id = p.role_id` + userRoleFilter +
```
[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/accesscontrol/database/database.go]: I checked by running the benchmark tests, and you're right - the performance doesn't change. I'll go for the simpler query with filtering by user ID for the whole query 😄

**Pull request merged**